rust/extns/numaflow-kafka/src/lib.rs

```diff
@@ -0,0 +1,680 @@
+pub enum KafkaAuth {
+    Sasl {
+        mechanism: String,
+        username: String,
+        password: String,
+    },
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TlsConfig {
+    pub insecure_skip_verify: bool,
+    pub ca_cert: Option<String>,
+    pub client_auth: Option<TlsClientAuthCerts>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TlsClientAuthCerts {
+    pub client_cert: String,
+    pub client_cert_private_key: String,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct KafkaSourceConfig {
+    pub brokers: Vec<String>,
+    pub topic: String,
+    pub consumer_group: String,
+    pub auth: Option<KafkaAuth>,
+    pub tls: Option<TlsConfig>,
+}
+
+/// Message represents a message received from Kafka which can be converted to Numaflow Message.
+#[derive(Debug)]
+pub struct KafkaMessage {
+    /// The user payload.
+    pub value: Bytes,
+    /// The partition number.
+    pub partition: i32,
+    /// The offset of the message.
+    pub offset: i64,
+    pub headers: HashMap<String, String>,
+}
+
+// A context can be used to change the behavior of consumers by adding callbacks
+// that will be executed by librdkafka.
+struct KafkaContext;
+
+impl ClientContext for KafkaContext {}
+
+impl ConsumerContext for KafkaContext {
+    fn pre_rebalance(&self, _: &BaseConsumer<Self>, rebalance: &Rebalance<'_>) {
+        info!("Pre rebalance {:?}", rebalance);
+    }
+
+    fn post_rebalance(&self, _: &BaseConsumer<Self>, rebalance: &Rebalance<'_>) {
+        info!("Post rebalance {:?}", rebalance);
+    }
+
+    fn commit_callback(&self, result: KafkaResult<()>, _offsets: &TopicPartitionList) {
+        info!("Committing offsets: {:?}", result);
+    }
+}
+
+type LoggingConsumer = StreamConsumer<KafkaContext>;
+
+enum KafkaActorMessage {
+    Read {
+        respond_to: oneshot::Sender<Result<Vec<KafkaMessage>>>,
+    },
+    Ack {
+        offsets: Vec<(i32, i64)>,
+        respond_to: oneshot::Sender<Result<()>>,
+    },
+    Pending {
+        respond_to: oneshot::Sender<Result<Option<usize>>>,
+    },
+}
+
+struct KafkaActor {
+    consumer: LoggingConsumer,
+    read_timeout: Duration,
+    batch_size: usize,
+    topic: String,
+    handler_rx: mpsc::Receiver<KafkaActorMessage>,
+}
+
+impl KafkaActor {
+    async fn start(
+        config: KafkaSourceConfig,
+        batch_size: usize,
+        read_timeout: Duration,
+        handler_rx: mpsc::Receiver<KafkaActorMessage>,
+    ) -> Result<()> {
+        let mut client_config = ClientConfig::new();
+        // https://docs.confluent.io/platform/current/clients/librdkafka/html/md_CONFIGURATION.html
+        client_config
+            .set("group.id", &config.consumer_group)
+            .set("bootstrap.servers", config.brokers.join(","))
+            .set("enable.partition.eof", "false")
+            .set("session.timeout.ms", "6000")
+            .set("enable.auto.commit", "false")
+            .set("auto.offset.reset", "earliest") // TODO: Make this configurable
```

Comment:
kafka supports about 300 different configs, how will be support those?

Reply:
Parsing the `config` field in the [kafka source](https://numaflow.numaproj.io/user-guide/sources/kafka/#option-2-built-in-kafka-source) section is not implemented yet. I'm thinking of doing that in a separate PR.

Reply:
There's option to set any number of configs https://docs.rs/rdkafka/0.37.0/rdkafka/config/struct.ClientConfig.html#impl-FromIterator%3C(String,+String)%3E-for-ClientConfig

Reply:
https://github.com/numaproj/numaflow/issues/2643